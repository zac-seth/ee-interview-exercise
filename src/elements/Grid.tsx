import { FC } from "react";

import { LayoutDirection } from "./layout-direction";
import { Layout } from "./Layout.style";

type GridPropsWithCols = { cols: number };
type GridPropsWithItems = {
  layoutDirection?: LayoutDirection;
  items: number[];
};

export type GridProps = GridPropsWithCols | GridPropsWithItems;

const hasCols = (props: GridProps): props is GridPropsWithCols =>
  !!Object.keys(props).find(key => key === `cols`);

export const Grid: FC<GridProps> = ({ children, ...props }) => {
  if (hasCols(props)) {
    const { cols } = props;

    return (
      <Layout
        layoutDirection={LayoutDirection.ROW}
        length={cols}
        data-testid="grid-layout"
      >
        {children}
      </Layout>
    );
  }

  const { layoutDirection = LayoutDirection.ROW, items } = props;

  return (
    <Layout
      layoutDirection={layoutDirection}
      items={items}
      data-testid="grid-layout"
    >
      {children}
    </Layout>
  );
};
