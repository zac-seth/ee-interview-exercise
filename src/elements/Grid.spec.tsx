import { render, screen } from '@testing-library/react';

import { Grid } from './Grid';

const COLS = 3;

const makeChildren = (length: number): JSX.Element[] =>
  Array.from({ length }, (_v, i) => (
    <div data-testid={`item-${i}`} key={`key-${i}`}>
      value {i}
    </div>
  ));

const makeGrid = () => <Grid cols={COLS}>{makeChildren(9)}</Grid>;

describe(`Grid`, () => {
  it(`Should render the grid with the list of given child items`, () => {
    render(makeGrid());

    expect(screen.getByTestId(`grid-layout`)).toMatchSnapshot();
  });

  it(`Should render the children in the grid`, async () => {
    render(makeGrid());

    const children = await screen.findAllByTestId(`item`, { exact: false });

    expect(children.length).toBe(9);
  });
});

