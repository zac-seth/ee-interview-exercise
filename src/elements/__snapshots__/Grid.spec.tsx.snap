// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid Should render the grid with the list of given child items 1`] = `
.c0 {
  display: grid;
  grid-gap: 0.25rem;
  margin: 0;
  padding: 0;
  grid-template-columns: repeat(3,1fr);
}

<div
  class="c0"
  data-testid="grid-layout"
>
  <div
    data-testid="item-0"
  >
    value 
    0
  </div>
  <div
    data-testid="item-1"
  >
    value 
    1
  </div>
  <div
    data-testid="item-2"
  >
    value 
    2
  </div>
  <div
    data-testid="item-3"
  >
    value 
    3
  </div>
  <div
    data-testid="item-4"
  >
    value 
    4
  </div>
  <div
    data-testid="item-5"
  >
    value 
    5
  </div>
  <div
    data-testid="item-6"
  >
    value 
    6
  </div>
  <div
    data-testid="item-7"
  >
    value 
    7
  </div>
  <div
    data-testid="item-8"
  >
    value 
    8
  </div>
</div>
`;
